torch_geometric/nn/aggr/basic.py

```diff
@@ -112,6 +112,150 @@ def forward(self, x: Tensor, index: Optional[Tensor] = None,
         return torch.sqrt(var.relu() + 1e-5)
 
 
+class QuantileAggregation(Aggregation):
+    r"""An aggregation operator that returns the feature-wise :math:`q`-th
+    quantile of a set :math:`\mathcal{X}` of :math:`n` elements. That is,
+    for every feature :math:`d`, computes
+
+    .. math::
+        \big[\mathrm{Q}_q(\mathcal{X})\big]_d = \begin{cases}
+            \mathbf{x}_{\pi_i,d} & i = q\cdot n, \\
+            f(\mathbf{x}_{\pi_i,d}, \mathbf{x}_{\pi_{i+1},d}) &
+                i < q\cdot n < i + 1,\\
+        \end{cases}
+
+    where :math:`\mathbf{x}_{\pi_1,d} \le \dots \le \mathbf{x}_{\pi_i,d} \le
+    \dots \le \mathbf{x}_{\pi_n,d}` and :math:`f(a, b)` is an interpolation
+    function defined by :obj:`interpolation`.
+
+    Args:
+        q (float): The quantile value :math:`q`. Must be a scalar in the range
+            :math:`[0, 1]`.
+        interpolation (str): Interpolation method applied if the quantile point
+            :math:`q\cdot n` lies between two values
+            :math:`a \le b`. Can be one of the following:
+
+            * :obj:`"lower"`: Returns the one with lowest value.
+
+            * :obj:`"higher"`: Returns the one with highest value.
+
+            * :obj:`"midpoint"`: Returns the average of the two values.
+
+            * :obj:`"nearest"`: Returns the one whose index is nearest to the
+              quantile point.
+
+            * :obj:`"linear"` (default): Returns a linear combination of the
+              two elments, defined as
+              :math:`f(a, b) = a + (b - a)\cdot(q\cdot n - i)`.
+
+        fill_value (float): Default value in the case no entry is
+            found for a given index (default: :obj:`NaN`).
+    """
+
+    interpolations = {'linear', 'lower', 'higher', 'nearest', 'midpoint'}
+
+    def __init__(self, q: float, interpolation: str = 'linear',
+                 fill_value: float = float('nan')):
+        if q > 1. or q < 0:
+            raise ValueError("'q' must be in the range [0, 1].")
+        if interpolation not in self.interpolations:
+            raise ValueError(f"Invalid '{interpolation}' interpolation method."
+                             f" Available interpolations:"
+                             f"{self.interpolations}")
+
+        super().__init__()
+        self.q = q
+        self.interpolation = interpolation
+        self.fill_value = fill_value
+
+    def forward(self, x: Tensor, index: Optional[Tensor] = None,
+                ptr: Optional[Tensor] = None, dim_size: Optional[int] = None,
+                dim: int = -2) -> Tensor:
+        self.assert_index_present(index)
+        assert index is not None  # required for TorchScript
+
+        # compute the quantile points
+        if dim_size is None:
+            dim_size = 0  # default value for `bincount`
```

Comment:
These assertion/if seem unavoidable for TorchScript (even if `self.assert_index_present(index)` already checks for `index`)

Reply:
Can you give a bit more detail on where you're seeing the problem?

Reply:
Sure. This is what happens without the first assertion:
```
RuntimeError: 

aten::bincount(Tensor self, Tensor? weights=None, int minlength=0) -> (Tensor):
Expected a value of type 'Tensor' for argument 'self' but instead found type 'Optional[Tensor]'.
:
  File "/opt/Workspace/pytorch_geometric/torch_geometric/nn/aggr/basic.py", line 180
        # if dim_size is None:
        #     dim_size = 0  # default value for `bincount`
        count = torch.bincount(index, minlength=dim_size)
                ~~~~~~~~~~~~~~ <--- HERE
        cumsum = torch.cumsum(count, dim=0)
        q_point = self.q * (count - 1) + cumsum - count
```
and this instead without the `if` statement:
```
RuntimeError: 

aten::bincount(Tensor self, Tensor? weights=None, int minlength=0) -> (Tensor):
Expected a value of type 'int' for argument 'minlength' but instead found type 'Optional[int]'.
:
  File "/opt/Workspace/pytorch_geometric/torch_geometric/nn/aggr/basic.py", line 180
        # if dim_size is None:
        #     dim_size = 0  # default value for `bincount`
        count = torch.bincount(index, minlength=dim_size)
                ~~~~~~~~~~~~~~ <--- HERE
        cumsum = torch.cumsum(count, dim=0)
        q_point = self.q * (count - 1) + cumsum - count
```
All the above were run with `torch.__version__ == 1.11.0`.

By the way, the same assertion is also present in the other aggregations, as it is done by [`self.reduce`](https://github.com/pyg-team/pytorch_geometric/blob/e805f0ca606ad67564d1c2de9e4a893243ca5d78/torch_geometric/nn/aggr/base.py#L112), while the `dim_size` check is done by [`scatter`](https://github.com/rusty1s/pytorch_scatter/blob/fc1b13945c8021c446aabd7b9207286b2239a899/torch_scatter/scatter.py#L76), since it also has an `Optional[int]` type hint.

Reply:
Sorry I missed your reply, I will take a look in detail tomorow

Reply:
Hmm yeah I see the issue... I can't see a good workaround for this right now other than including the check as you have here, but we're going to have this with all `Aggregation` modules.

I think lets leave the assert here and maybe we can add an issue to address this later? WDYT?

cc @rusty1s

Reply:
Summay of my understanding:
- issue 1: `jit` by default looks ad 'forward', and we'd want it to look at '__call__'
- issue 2: `jit` cannot handle `**kwargs` that we have in `__call__`
- issue 3: `jit` cannot handle `super().` that we have in `__call__`

Reply:
@Padarn I agree, this should become a separate issue, as it may lead to some refactoring (maybe the `assert_*_present` can forward/return `index`/`ptr`/etc? This should solve the `Optional[Tensor]` issue, but not the `Optional[int]` one of `dim_size`).